### ISSUE TITLE
Multi-label topics not captured

### ISSUE THREAD

**Issue by cassinthangam4996** [huggingface/setfit#316]
I am working on a task to classify a sentence into multiple topics (Multi-label classification). 
Initially, I trained the entire 1.5M unlabeled data on bert base model using domain adaptation.
Then I manually annotated the labels with atleast 10 examples each and fine-tuned it.
The test accuracy what I received was 35% for 1 epoch and 46% for 7 epochs.
I see that the model is capturing negations too but it is not capturing multi-labels. 
I know that there is no way to measure accuracy for train/test dataset after every epoch to get to know if it is overfitting or underfitting.
Can you suggest some some good range of epochs for this multi label classification or changes in hyper-parameter opt. settings? Or may be i should increase the quality of the labeled sentences? Suggestions please.

**Comment by tomaarsen** [comment]
Hello @cassin-edwin!

By "not captured", are you referring to a lower-than-expected performance or a bug? I'll assume the former for the rest of this message. 
How many classes (topics) are you trying to train on? I do not have a lot of experience with multi-label classification, but with single-label classification, the number of classes has a notable impact on the number of training samples that should be annotated for solid performance. The exact details behind these graphs are irrelevant, but they do show some interesting behaviour. Note that the boxplots show the performances of three very similar SetFit models:

## Performance of k-shot on the [`bbc-news`](https://huggingface.co/datasets/SetFit/bbc-news) dataset
**Important**: This dataset has 2 classes!
![bbc-news](https://user-images.githubusercontent.com/37621491/220567775-e9f5abbf-7a9e-403a-a07e-627d5b9423d1.png)
This is a fairly simple dataset, and SetFit is able to reach 90%+ accuracies with only 4 samples per class (e.g. 8 annotated samples total)

## Performance of k-shot on the [`emotion`](https://huggingface.co/datasets/dair-ai/emotion) dataset
**Important**: This dataset has 6 classes!
![emotion](https://user-images.githubusercontent.com/37621491/220568982-0b20f9b0-84f8-4067-aa5b-8377a388d984.png)
This is a fairly difficult dataset, even for humans. SetFit will require more data before it reaches a point where additional data stops improving the performance. This is caused both by how there are more classes that need to be separated in the embedding space, and because of the relatively low data quality.

The takeaway here is that adding more data may be very fruitful for models that classify between more classes. For your specific case, I am going to guess that it will help, especially considering most models that I train do not improve beyond epoch 1: the training loss has generally already reached 0 then (although the evaluation loss is not 0).

Another interesting test for your situation is to take a "general" pretrained sentence transformer, rather than your domain adapted one. It would be interesting to see the performance differences.

As for hyperparameter tuning, I've been trying to find some better values myself, but the only improvement that I've found for my scenarios is that a learning rate of `8e-5` instead of `2e-5` sometimes works better, and that a slightly higher `num_iterations` (e.g. 30) can help, too.
However, I think the main thing to invest time in is getting some additional annotations.

- Tom Aarsen

**Comment by tomaarsen** [comment]
It is common for SetFit models to perform best at or near 1 epoch. If you want to do more training, it is recommended to increase `num_iterations` instead, which impacts the number of contrastive training pairs that are generated internally.
For non-multi label datasets it is possible to perform methods (PCA, t-SNE) to decrease the embedding dimensions of your evaluation set down to 3, resulting in graphs like:
![emotion_64_shot_20_num_iter_large](https://user-images.githubusercontent.com/37621491/220842287-8f22d6fa-d697-492d-afe9-50ca8cedf574.png)
However, I'm not sure if this can reasonably be done with multi-label data. 

For data labeling, perhaps you'll enjoy using Argila with SetFit, there's a tutorial [here](https://docs.argilla.io/en/latest/tutorials/notebooks/training-textclassification-setfit-sentiment.html). Note: I am affiliated with Argilla.

- Tom Aarsen

**Comment by vahuja4** [comment]
@cassin-edwin - could you please tell how you did the domain adpatation? And, why do you think that the domain adapted model is doing worse than the original st?
